Added fallback time extraction engine

TheMulti0 opened the pull request:
Solves #134, where for some reason time is not always correctly extracted, therefore I have added a fallback way of extracting time (if the current one fails), by trying to extract it from the element's text.

As of right now, it successfully works with two different type of dates:
Example 1: `October 28 11:58 AM`
Example 2: `9 hrs` (9 hours ago, the example was only tested with hours)

The engine uses the following regex:
`(\w+ \d{2} at \d{2}:\d{2} (AM|PM))|(\d{1,2} \w+)`
with a segment for example 1 (`\w+ \d{2} at \d{2}:\d{2} (AM|PM)`) and for example 2 (`(\d{1,2} \w+`).

In order to easily parse dates, I have added a dependency on the package `dateparser` (added to `requirements.txt` and `requirements-dev.txt`)

Even though not all example of dates have been tested yet, the regex will probably be able to extract the different undiscovered date templates and let the date parser parse it automatically.

balazssandor commented:
Thanks!

balazssandor reviewed (commented):
Doesn't work in cases:
```
-> 'time': dateparser.parse(time)
(Pdb) time
'13 at'
(Pdb) dateparser.parse(time)
datetime.datetime(2020, 11, 13, 0, 0)
```
dateparser is not able to find the date from `13 at` string 

This was the date appearance from where it failed:
![image](https://user-images.githubusercontent.com/32865690/98236035-60260c00-1f6b-11eb-84ce-fe170c2b42d7.png)

balazssandor commented on facebook_scraper/extractors.py:
```diff
@@ -203,6 +206,16 @@ def extract_time(self) -> PartialPost:
+            if time_match:
+                time = time_match.group(0)
+                return {
+                    'time': dateparser.parse(time)
```
https://github.com/kevinzg/facebook-scraper/pull/135#pullrequestreview-524168255

TheMulti0 replied:
I'm not sure I understood the problem, you tried to scrap a post with the date 'October 13 at 9:44 PM', or did you send the string 'at 13' to dateparser?

I would love seeing the original post if so.

balazssandor replied:
The thing is that the matched group was `ASFOROctober 13 at 8:44 PM`, including a part of the page name or title.
I came with this ugly regex fix
`time_regex = re.compile(r"((Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:tember)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?) \d{1,2} at \d{1,2}:\d{2} (AM|PM))|(\d{1,2} \w+)")` 

and that's how `(\d{1,2} \w+)` ended up finding `13 at`

balazssandor replied:
Added also regex for `Yesterday at 12:30 PM`-like matching also:
`((Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:tember)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?) \d{1,2} at \d{1,2}:\d{2} (AM|PM))|Yesterday at \d{1,2}:\d{2}|(\d{1,2} \w+)`

evoup replied:
Thks balazssandor, there is also has Today, so will be this:
((Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:tember)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?) \d{1,2} at \d{1,2}:\d{2} (AM|PM))|Today at \d{1,2}:\d{2} (AM|PM)|Yesterday at \d{1,2}:\d{2} (AM|PM)|(\d{1,2} \w+)

TheMulti0 commented:
@evoup @balazssandor Thanks for the help!

I added the new regex to `utils.py`, in the latest commit (also extracted it a little bit to make it more readable), mind checking it out?

valterartur commented:
Dateparser is not threadsafe, so I think it's not a really good realization. This [issue](https://github.com/scrapinghub/dateparser/issues/441) is still open on GitHub. I have got some errors with your way.

TheMulti0 commented:
Thanks for the issue, I will implement custom date parsing with the desired formats.

> I have got the same errors with your way.

What errors did you get with 'my way'? Thread safe errors?

valterartur commented:
> Thanks for the issue, I will implement custom date parsing with the desired formats.
> 
> > I have got the same errors with your way.
> 
> What errors did you get with 'my way'? Thread safe errors?

I have got a problem described in the referenced issue ticket.

baraths92 commented:
> @evoup @balazssandor Thanks for the help!
> 
> I added the new regex to `utils.py`, in the latest commit (also extracted it a little bit to make it more readable), mind checking it out?

Thank you. I tried it and it worked. Thanks a lot once again. The original author should consider fixing this in his code.

TheMulti0 commented:
There are multiple different date formats we need to parse, since as @valterartur stated, using `dateparser` is not trival.

1. `Oct 16 at 11:00 PM`
2. `October 16 at 11:00 PM`
3. `Yesterday at 11:00 PM`
4. `Today at 11:00 PM`
5. `16 hrs` (16 hours ago)
6. `16h` (16 hours ago)

richardscollin commented on requirements.txt:
```diff
@@ -92,3 +92,7 @@ websockets==8.0.2 \
     --hash=sha256:049e694abe33f8a1d99969fee7bfc0ae6761f7fd5f297c58ea933b27dd6805f2 \
     --hash=sha256:882a7266fa867a2ebb2c0baaa0f9159cabf131cf18c1b4270d79ad42f9208dc5
+
+html2text~=2020.1.16
+requests~=2.24.0
+dateparser~=1.0.0
```
Not a maintainer, but I used your version as a temporary work around. There was an issue in the requirements.txt file you pushed. There's a flag that requires you include the hash. Also you probably should order them alphabetically in the file.

```
html2text~=2020.1.16 \
    --hash=sha256=c7c629882da0cf377d66f073329ccf34a12ed2adf0169b9285ae4e63ef54c82b 
```

```
dateparser~=1.0.0 \
    --hash=sha256=17202df32c7a36e773136ff353aa3767e987f8b3e27374c39fd21a30a803d6f8 
```

kevinzg replied:
This file can be generated with poetry, if you don't use poetry I can update it after merging.

kevinzg commented on facebook_scraper/extractors.py:
```diff
@@ -203,6 +203,15 @@ def extract_time(self) -> PartialPost:
+                return {
+                    'time': date
+                }
+        except:
```
It's better to catch `Exception`:
```suggestion
        except Exception:
```
Or maybe let it fail since it should be caught and logged on the extract_post method.

TheMulti0 replied:
You're right, I'll let it fall. Should be caught later on.

kevinzg commented:
I think this is alright.
The only issue seems to be that it isn't thread safe, however anyone that needs a thread safe date parsing function could monkeypatch `utils.parse_date`, so I'm inclined to merge it with the dateparser library unless someone suggests a better option.

TheMulti0 commented:
> I think this is alright.
> The only issue seems to be that it isn't thread safe, however anyone that needs a thread safe date parsing function could monkeypatch `utils.parse_date`, so I'm inclined to merge it with the dateparser library unless someone suggests a better option.

I agree, this change is a fallback solution and the benefits are bigger than the downsides (the thread safety bug).
Parsing for all of the different 5 formats I specified isn't extremely complicated, it can be implemented if there is a need.

Let's merge!

TheMulti0 commented:
Hi, can we please merge?

kevinzg commented on facebook_scraper/utils.py:
```diff
@@ -43,3 +46,36 @@ def make_html_element(html: str, url=DEFAULT_URL) -> Element:
+        r"Dec(?:ember)?|" \
+        r"Yesterday|" \
+        r"Today"
+date = f"({month}) " + r"\d{1,2}"
```
Why are you capturing the month?
Some dates include the year, like: `April 3, 2018 at 8:02 PM`
```suggestion
date = f"(?:{month}) " + r"\d{1,2}" + r"(?:, \d{4})?"
```

kevinzg commented on facebook_scraper/utils.py:
```diff
@@ -43,3 +46,36 @@ def make_html_element(html: str, url=DEFAULT_URL) -> Element:
+hour = r"\d{1,2}"
+minute = r"\d{2}"
+period = r"AM|PM"
+exact_time = f"({date}) at {hour}:{minute} ({period})"
```
Again, I think it shouldn't capture the date or period.
```suggestion
exact_time = f"(?:{date}) at {hour}:{minute} (?:{period})"
```

TheMulti0 replied:
I'm not sure I understand, why again shouldn't the whole date and am\pm be captured?

kevinzg replied:
Because those groups are not being used.

In `parse_datetime` only the group 0 is being retrieved, which is whole matched string.
It would make sense to capture every element of the date separately if we were parsing the date ourselves, but we are just passing the whole string to dateparser.

TheMulti0 replied:
I see. Thanks!

kevinzg commented on facebook_scraper/utils.py:
```diff
@@ -43,3 +46,36 @@ def make_html_element(html: str, url=DEFAULT_URL) -> Element:
+minute = r"\d{2}"
+period = r"AM|PM"
+exact_time = f"({date}) at {hour}:{minute} ({period})"
+relative_time = r"\d{1,2} \w+"
```
This seems a bit vague since it is being matched with the text from the whole post and not just a date.
What exactly is this trying to match?

kevinzg replied:
According to a previous comment it could be 16 hrs or 16h so maybe:
```suggestion
relative_time = r"\b\d{1,2}(?:h| hrs)"
```

TheMulti0 replied:
The only date I found is in the entire post text, and I only saw dates that have explicit month and day of month specification and a time, days that have 'Today' or 'Yesterday' and a time, and just relative time dates (16 hours ago, hours is the only case I found)

kevinzg commented on facebook_scraper/extractors.py:
```diff
@@ -203,6 +203,12 @@ def extract_time(self) -> PartialPost:
             except (KeyError, ValueError):
                 continue
 
+        date = utils.parse_datetime(element_full_text=self.element.full_text)
```
How complicate would it be to select just the html element with the date?

TheMulti0 replied:
I do not understand your suggestion

kevinzg replied:
I mean using `self.element.find('abbr', first=True).text` instead of `self.element.full_text`, since the date seems to always be inside an `abbr` tag.

kevinzg replied:
I'll add that change and merge it.
